### ISSUE TITLE
UnusedPrivateMember complains about range loop variable

### ISSUE THREAD

**Issue by theHacker** [detekt/detekt#5816]
We are experiencing a false-positive for `UnusedPrivateMember`. I found https://github.com/detekt/detekt/issues/4879 but it's closed without any comment. Is the rule's behavior intended or is it a bug?

I am not sure, if you like to reopen old issues, so I just create a fresh one to discuss.

## Expected Behavior

```kt
for (i in 1..someVar) {
    doSomething()
}
```
should not report any issues

## Observed Behavior

Detekt complains
```
Private property `i` is unused. [UnusedPrivateMember]
```

## Context

We could rewrite with `forEach()` or `repeat()`, but all these solutions are calling internally only our initial code, which I assume is the best code.
```kt
// workarounds

(1..someVar).forEach {
    doSomething()
}

repeat(someVar) {
    doSomething()
}
```

Another nicer workaround would be to add `i` into the rule's `allowedNames`.

## Your Environment
* Version of detekt used: 1.21.0
* Version of Gradle used (if applicable): 7.6
* Operating System and version: locally MacOS 12.6.1, CI with GitHub Actions

**Comment by BraisGabin** [comment]
For me this is a place where `repeat` should be used. Why do you think it is a workaround instead of a fix?

**Comment by cortinico** [comment]
```kotlin
for (i in 1..someVar) {
    doSomething()
}
```

As @BraisGabin said, this is a scenario for a `repeat()` or you can use this other notation which won't trigger a `UnusedPrivateMember`:

```kotlin
(1..someVar).forEach {
    doSomething()
}
```

**Comment by theHacker** [comment]
For `forEach` the same argument applies.

You can easily check why I call that a "workaround" when you compare the generated byte code of both solutions.
The normal loop just counts, while for lambda variant has to construct the range and calls its iterator. 

![image](https://user-images.githubusercontent.com/7399011/221560105-8e6ab70c-1028-479b-bbd4-e4e06515cfcc.png)

<details>
   <summary>Code used to generate</summary>

```kt
package foo

object Test {

    private val someVar = 5

    fun normalLoop() {
        for (i in 1..someVar) {
            print(i)
        }
    }

    fun lambda() {
        (1..someVar).forEach {
            print(it)
        }
    }

    fun repeat() {
        repeat(someVar) {
            print(it)
        }
    }
}
```

</details>

It's a huge difference. Yes, we talking micro optimization, I know that. Nobody would notice any difference in the end.
However, to me this is a principle thing. I teach young devs the differences of both solutions and why one should prefer one over the other. It's not beneficial then when a linter is working against the better solution.

---

I imagine it is hard to decide when it's a dead variable (linter should warn) vs. when it's a legit loop (linter should not warn).
Maybe just not raise an issue when you see it's a range of numbers, i.e. `(U)IntRange` or `(U)LongRange`?

**Comment by cortinico** [comment]
Quoting a previous comment of mine: https://github.com/detekt/detekt/issues/3823#issuecomment-959819563

> I recall seeing similar issues in the past, mentioning minor tweaks to get some perf. Improvements. I believe this is going a bit in an opposite direction than Detekt's mission: helping you write clean code (the cleaner code might not necessarily be the most performant).

I believe we should strive for cleaner code. In your case you should probably disable `UnusedPrivateMember` and implement your own version of it? Or perhaps use the `allowedNames` to specify a name you'll be using for iterators.

**Comment by theHacker** [comment]
I see your point 😄 

---

For the concrete case why I reported we solved the issue by using `repeat()` which was semantically fitting.
So if you don't intend to change the rule, feel free to close the issue.

**Comment by blundell** [comment]
careful 

![Image](https://github.com/user-attachments/assets/8b77cb76-2eb7-400e-894d-d748bdf680e9)

**Comment by BraisGabin** [comment]
You are not using `i`. This is not a false positive. For that you can add +1 or range should have a size, or count. You can use that inside the repeat. There are multiple ways to do the same without a `for`.
